plan.go

```diff
@@ -0,0 +1,206 @@
+
+		cursor := bucket.Cursor()
+		for ref, bytes := cursor.First(); bytes != nil; ref, bytes = cursor.Next() {
+			if err := scanner(ref, bytes); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func (i indexScanQueryPlan) execute(walker ReferenceScanFn) error {
+	queryParts, err := i.index.QueryPartsOutsideIndex(i.query)
+	if err != nil {
+		return err
+	}
+	if len(queryParts) != 0 {
+		return errors.New("no index with exact match to query found")
+	}
+
+	// do the IndexScan
+	return i.collection.db.View(func(tx *bbolt.Tx) error {
+		// nil is not possible since adding an index creates the iBucket
+		iBucket := tx.Bucket([]byte(i.collection.Name))
+		if iBucket == nil { // nothing added yet
+			return nil
+		}
+
+		// expander expands the index entry to the actual document
+		expander := indexEntryExpander(walker)
+
+		return i.index.Iterate(iBucket, i.query, expander)
+	})
+}
+
+func (i resultScanQueryPlan) execute(walker DocWalker) error {
+	queryParts, err := i.index.QueryPartsOutsideIndex(i.query)
+	if err != nil {
+		return err
+	}
+
+	// do the IndexScan
+	return i.collection.db.View(func(tx *bbolt.Tx) error {
+		docBucket := i.collection.documentBucket(tx)
+		if docBucket == nil {
+			// no bucket means no docs
+			return nil
+		}
+
+		// nil is not possible since adding an index creates the iBucket
+		iBucket := tx.Bucket([]byte(i.collection.Name))
+
+		// resultScanner takes the refs from the indexScan, resolves the document and applies the remaining queryParts
+		resultScan := resultScanner(queryParts, walker)
+
+		// fetcher expands references to documents, for each document it calls the resultScan
+		fetcher := documentFetcher(docBucket, resultScan)
+
+		// expander expands the index entry to the actual document
+		expander := indexEntryExpander(fetcher)
+
+		return i.index.Iterate(iBucket, i.query, expander)
+	})
+}
+
+// documentFetcher creates a ReferenceScanFn which is called with a reference, fetches the document and calls the documentScanFn
+func documentFetcher(documentCollection *bbolt.Bucket, docWalker documentScanFn) ReferenceScanFn {
+	return func(key []byte, ref []byte) error {
+		if documentCollection == nil {
+			return nil
+		}
+		docBytes := documentCollection.Get(ref)
+		if docBytes != nil {
+			return docWalker(ref, docBytes)
+		}
+		return nil
+	}
+
+}
+
+// resultScanner returns a resultScannerFn. For each call it will compare the document against the given queryParts.
+// If conditions are met, it'll call the DocWalker
+func resultScanner(queryParts []QueryPart, walker DocWalker) documentScanFn {
+	return func(ref []byte, docBytes []byte) error {
+		doc := DocumentFromBytes(docBytes)
+	outer:
+		for _, part := range queryParts {
+			keys, err := doc.KeysAtPath(part.Name())
+			if err != nil {
+				return err
+			}
+			for _, k := range keys {
+				m, err := part.Condition(k, nil)
+				if err != nil {
+					return err
+				}
+				if m {
+					continue outer
```

Comment:
isn't `return nil` simpler if `m` is false, because it doesn't require labels/goto?

Reply:
at least one has to match, not all. As soon as 1 is found, the search can continue with the next part.